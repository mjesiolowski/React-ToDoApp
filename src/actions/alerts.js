export const lengthAlert = isTrue => ({
   type: 'LENGTH_ALERT',
   isTrue
})

export const duplicateAlert = isTrue => ({
   type: 'DUPLICATE_ALERT',
   isTrue
})

// export const editAlert = isTrue => ({
//    type: 'EDIT_ALERT',
//    isTrue
// })

// export const editionActive = isTrue => ({
//    type: 'EDITION_ACTIVE',
//    isTrue
// })

// export const searchAlert = isTrue => ({
//    type: 'SEARCH_ALERT',
//    isTrue
// })