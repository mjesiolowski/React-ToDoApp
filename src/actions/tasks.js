import uuid from 'uuid'
import moment from 'moment'

export const addTask = (
   name = "default task",
   active = true,
   createdAt = moment().valueOf(),
   isPriority = false,
   comments = []
) => ({
   type: "ADD_TASK",
   task: {
      id: uuid(),
      name: name.trim(),
      active,
      createdAt,
      isPriority,
      comments
   }
})

export const editTask = (id, update) => ({
   type: "EDIT_TASK",
   id,
   update,
})

export const removeTask = (id) => ({
   type: "REMOVE_TASK",
   id
})

export const doneTask = (id) => ({
   type: "DONE_TASK",
   id
})

export const addComment = (id, text) => ({
   type: "ADD_COMMENT",
   id,
   comment: {
      id: uuid(),
      text,
      createdAt: moment().valueOf(),
      beingEdited: false
   }
})

export const editComment = (id, update) => ({
   type: "EDIT_COMMENT",
   id,
   update,
})

export const deleteComment = (taskId, commentId) => ({
   type: "DELETE_COMMENT",
   taskId,
   commentId
})