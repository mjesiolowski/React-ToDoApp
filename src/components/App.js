import React, { Component } from 'react';
import Header from './Header'
import TasksList from './TasksList'
import uuid from 'uuid'


class App extends Component {
  state = {
    editedInputValue: '',
    headerInputValue: '',
    minimalTaskLengthAlert: false,
    searchSectionActive: false,
    searchedTasks: [],
    taskDuplicatedAlert: false,
    tasks: [
      {
        id: uuid(),
        text: "Task 1",
        active: true,
        edited: false,
      },
      {
        id: uuid(),
        text: "Task 2",
        active: true,
        edited: false,
      },
      {
        id: uuid(),
        text: "Task 3",
        active: true,
        edited: false,
      },
    ],
  }

  //INPUT HANDLERS

  handleHeaderInputValue = e => {
    this.setState({
      headerInputValue: e.target.value
    })
  }

  handleHeaderInputSubmit = e => {
    e.preventDefault()
    this.handleAddTaskButton()
  }

  handleEditInputValue = e => {
    this.setState(({
      editedInputValue: e.target.value
    }))
  }

  //ADDING TASK
  // handleAddTaskButton = () => {
  //   const tasksToDo = [...this.state.tasks]
  //   const duplicateCheck = tasksToDo.filter(task => task.text === this.state.headerInputValue && task.active).length

  //   if (this.state.headerInputValue.length <= 2) {
  //     this.setState({
  //       minimalTaskLengthAlert: true
  //     })
  //   }

  //   else if (duplicateCheck) {
  //     this.setState({
  //       taskDuplicatedAlert: true
  //     })
  //   }
  //   else {
  //     tasksToDo.push(
  //       {
  //         id: uuid(),
  //         text: this.state.headerInputValue,
  //         active: true,
  //         edited: false,
  //       }
  //     )
  //     this.setState({
  //       tasks: tasksToDo,
  //       headerInputValue: ''
  //     })
  //   }
  // }

  //CHANGING TASK STATUS
  handleDoneTaskButton = id => {
    const tasks = [...this.state.tasks]
    tasks.forEach(task => {
      if (task.id === id) {
        task.active = !task.active
        task.time = new Date().getTime()
      }
    })
    this.setState({
      tasks,
    })
  }

  //EDITING TASK
  handleEditButton = (id, task) => {
    if (!this.state.editedInputValue) {
      this.setState({
        editedInputValue: task
      })

      const tasks = [...this.state.tasks]
      tasks.forEach(task => {
        if (task.id === id) {
          task.edited = !task.edited
        }
      })
      this.setState({
        tasks,
      })
    }
  }

  handleEditingTask = (e, id) => {
    e.preventDefault()
    const tasks = [...this.state.tasks]

    if (this.state.editedInputValue.length < 3) { alert('Task too short') }
    else {
      tasks.map(task => {
        if (task.id === id) {
          task.text = this.state.editedInputValue
          task.edited = false
        } return null
      })

      this.setState(() => ({
        editedInputValue: '',
        tasks,
      }))
    }
  }

  //REMOVING TASK
  handleRemoveTaskButton = id => {
    const tasks = [...this.state.tasks]
    const tasksAfterRemoval = tasks.filter((task) => {
      if (task.id !== id) {
        return task
      } return null
    })

    this.setState({
      tasks: tasksAfterRemoval,
    })
  }

  //SEARCHING TASK
  handleSearchTaskButton = () => {
    const tasks = [...this.state.tasks]

    if (this.state.headerInputValue.length) {
      const searchResult = tasks.filter((result) => result.text.toLowerCase().includes(this.state.headerInputValue.trim().toLowerCase()) && result.active
      )

      this.setState({
        headerInputValue: '',
        searchSectionActive: true,
        searchedTasks: [...searchResult]
      })
    }
  }

  handleReturnButton = () => {
    this.setState({
      searchSectionActive: false,
    })
  }

  //ALERT HANDLERS

  componentDidUpdate(prevProps, prevState) {
    if (this.state.minimalTaskLengthAlert && this.state.headerInputValue.length > 2) {
      this.setState({
        minimalTaskLengthAlert: false
      })
    }
    else if (prevState.headerInputValue !== this.state.headerInputValue) {
      this.setState({
        taskDuplicatedAlert: false
      })
    }
  }

  //RENDERING

  render() {
    return (
      <div className="App">

        <header className="header">
          <Header
            alert={this.state.minimalTaskLengthAlert}
            duplicate={this.state.taskDuplicatedAlert}
            value={this.state.headerInputValue}

            handleAddTaskButton={this.handleAddTaskButton}
            handleHeaderInputSubmit={this.handleHeaderInputSubmit}
            handleHeaderInputValue={this.handleHeaderInputValue}
            handleSearchTaskButton={this.handleSearchTaskButton}
          />
        </header>

        <main className="tasksList">
          <TasksList
            editedValue={this.state.editedInputValue}
            searchSection={this.state.searchSectionActive}
            searchedTasks={this.state.searchedTasks}
            tasksToDo={this.state.tasks}

            handleEditButton={this.handleEditButton}
            handleEditInputValue={this.handleEditInputValue}
            handleEditingTask={this.handleEditingTask}
            handleDoneTaskButton={this.handleDoneTaskButton}
            handleReturnButton={this.handleReturnButton}
            handleRemoveTaskButton={this.handleRemoveTaskButton}
          />
        </main>

      </div>
    );
  }
}

export default App;
