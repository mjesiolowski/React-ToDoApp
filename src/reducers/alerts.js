const alertsReducerDefaultState = {
   lengthAlert: false,
   duplicateAlert: false,
   dateAlert: false,
   commentAlert: false,
}

export const alertsReducer = (state = alertsReducerDefaultState, action) => {
   switch (action.type) {
      case 'LENGTH_ALERT':
         return {
            ...state,
            lengthAlert: action.isTrue
         }

      case 'DUPLICATE_ALERT':
         return {
            ...state,
            duplicateAlert: action.isTrue
         }

      case 'DATE_ALERT':
         return {
            ...state,
            dateAlert: action.isTrue
         }

      case 'COMMENT_ALERT':
         return {
            ...state,
            commentAlert: action.isTrue
         }

      default:
         return state
   }
}