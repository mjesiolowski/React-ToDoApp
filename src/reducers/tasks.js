import uuid from 'uuid'
import moment from 'moment'

const tasksReducerDefaultState = [{
   id: '123',
   name: "Task 1",
   active: true,
   createdAt: moment().valueOf(),
   dueDate: moment().valueOf(),
   isPriority: false,
   comments: []

},
{
   id: uuid(),
   name: "Task 2",
   active: true,
   createdAt: moment().valueOf(),
   dueDate: moment().valueOf(),
   isPriority: false,
   comments: []
},
{
   id: uuid(),
   name: "Task 3",
   active: false,
   createdAt: moment().valueOf(),
   dueDate: moment().valueOf(),
   isPriority: false,
   comments: []
},
]

export const tasksReducer = (state = tasksReducerDefaultState, action) => {
   switch (action.type) {
      case 'ADD_TASK':
         return [
            ...state,
            action.task
         ]
      case 'EDIT_TASK':
         return state.map(task => {
            if (task.id === action.id) {
               return {
                  ...task,
                  ...action.update
               }
            } else {
               return task
            }
         })

      case 'DONE_TASK':
         return state.map(task => {
            if (task.id === action.id) {
               return {
                  ...task,
                  ...task.active = false,
                  ...task.time = new Date().getTime()
               }
            } else {
               return task
            }
         })
      case 'REMOVE_TASK':
         return state.filter(task => task.id !== action.id)

      case 'ADD_COMMENT':
         return state.map((task) => {
            if (task.id === action.id) return {
               ...task,
               comments: [
                  ...task.comments,
                  action.comment
               ]
            }
            else return task
         })


      case 'EDIT_COMMENT':
         return state.map((task) => {

            if (task.id === action.id) {
               const updatedComments = task.comments.slice().filter(comment => comment.id !== action.update.id)

               return {
                  ...task,
                  comments: [
                     ...updatedComments,
                     action.update
                  ]
               }
            }
            else return task
         })

      case 'DELETE_COMMENT':
         return state.map((task) => {

            if (task.id === action.taskId) {
               const updatedComments = task.comments.slice().filter(comment => comment.id !== action.commentId)

               return {
                  ...task,
                  comments: [
                     ...updatedComments
                  ]
               }
            }
            else return task
         })

      default:
         return state
   }
}