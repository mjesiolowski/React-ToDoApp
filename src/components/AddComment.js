import React, { useState } from 'react'
import { connect } from 'react-redux'
import { addComment } from '../actions/tasks'
import { commentAlert as commentAlertAction } from '../actions/alerts'

export const AddComment = ({ addComment, commentAlertAction, taskId, commentAlert }) => {

   const [commentValue, setCommentValue] = useState("")

   const handleSubmit = (e) => {
      e.preventDefault()
      if (commentValue.length > 0) {
         addComment(taskId, commentValue)
         commentAlertAction(false)
         setCommentValue("")
      }
      else {
         commentAlertAction(true)
      }
   }

   return (
      <form
         onSubmit={handleSubmit}
         className="form"
      >
         <input
            value={commentValue}
            onChange={(e) => setCommentValue(e.target.value)}
            type="text"
            placeholder="Add comment"
            className="form__input" />

         <button className="task__button button">Add comment</button>

         {commentAlert && <p className='text'>Minimum length of 1 character required</p>}
      </form>

   )
}

const mapDispatchToProps = (dispatch) => ({
   addComment: (taskId, commentValue) => dispatch(addComment(taskId, commentValue)),
   commentAlertAction: (isTrue) => dispatch(commentAlertAction(isTrue))
}
)
const mapStateToProps = ({ alerts }) => {
   return {
      commentAlert: alerts.commentAlert
   }
}

export default connect(mapStateToProps, mapDispatchToProps)(AddComment)