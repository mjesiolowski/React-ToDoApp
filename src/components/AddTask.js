import React, { useState } from 'react';
import { connect } from 'react-redux'
import { addTask } from '../actions/tasks'
import moment from 'moment'
import { validateTask } from '../functions/validateTask'
import { handleAlerts } from '../functions/handleAlerts'

const AddTask = ({ dispatch, alerts, tasks }) => {
   const dateFormat = "DD.MM.YYYY"

   const [taskName, setTaskName] = useState("")
   const [taskPriority, setTaskPriority] = useState(false)
   const [taskDeadline, setTaskDeadline] = useState(moment().format(dateFormat))

   const task = {
      taskName,
      taskDeadline,
      dateFormat
   }

   const handleAddTask = (e, dateFormat) => {
      e.preventDefault()
      const addTaskAction = () => dispatch(addTask({
         name: taskName,
         isPriority: taskPriority,
         deadline: moment(taskDeadline, dateFormat).valueOf()
      }))

      const isTaskValid = handleAlerts(validateTask({ ...task }, tasks), dispatch)
      if (isTaskValid) {
         addTaskAction()
         setTaskName("")
      }
   }

   return (
      <section className="addTask">
         <h2 className="title__sub addTask__title">AddTask</h2>
         <form
            onSubmit={(e) => handleAddTask(e, dateFormat)}
            className='form'
         >
            <input
               type="text"
               placeholder="Add task"
               value={taskName}
               onChange={(e) => setTaskName(e.target.value)}
               className='form__input'
            />

            <input
               value={taskDeadline}
               onChange={(e) => setTaskDeadline(e.target.value)}
               className='form__input'
            />

            <div className="form__wrapper">
               <label
                  htmlFor="priority"
                  className='form__label'
               >Priority: </label>

               <select
                  id="priority"
                  value={taskPriority}
                  onChange={(e) => setTaskPriority(e.target.value === "true" ? true : false)}
                  className='form__select'
               >
                  <option value="false">Not important</option>
                  <option value="true">Important</option>

               </select>
            </div>

            {alerts.duplicateAlert && <p className='addTask__alert text'>Task already on the to do list</p>}
            {alerts.lengthAlert && <p className='addTask__alert text'>Minimum 3 characters required</p>}
            {alerts.dateAlert && <p className='addTask__alert text'>Your deadline date must not be set in the past. Date fromat: DD.MM.YYYY
               </p>}
            <button className='addTask__button button'><i className="fas fa-plus-circle"></i></button>
         </form>
      </section>
   )
}

const mapStateToProps = ({ tasks, alerts }) => ({
   tasks,
   alerts
})

export default connect(mapStateToProps)(AddTask)