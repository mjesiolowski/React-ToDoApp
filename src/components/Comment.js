import React, { useState } from 'react';
import { connect } from 'react-redux'
import moment from 'moment'
import { editComment, deleteComment } from '../actions/tasks'

const Comment = ({ comment, dispatch, id: taskId, completed }) => {

   const [commentText, setCommentText] = useState(comment.text)

   const handleEditComment = () => {
      dispatch(editComment(taskId, { ...comment, beingEdited: true, }))
   }


   const handleDeleteComment = () => {
      dispatch(deleteComment(taskId, comment.id))
   }

   const handleSubmit = (e) => {
      e.preventDefault()
      dispatch(editComment(taskId,
         {
            ...comment,
            beingEdited: false,
            text: commentText
         }))
   }

   return (
      <>
         {comment.beingEdited ?
            <form
               onSubmit={handleSubmit}
               className="form"
            >
               <input
                  type="text"
                  value={commentText}
                  onChange={(e) => setCommentText(e.target.value)}
                  className="form__input"
               />
               <p className="text">{moment(comment.createdAt).format('DD-MM-YYYY')}</ p>
            </form>

            :

            <div className="task__comment">
               <p className="task__text text">{commentText}</p>
               <p className="task__text text">{moment(comment.createdAt).format('DD-MM-YYYY')}</ p>

               {!completed && <button
                  onClick={handleEditComment}
                  className="task__button button">Edit comment</button>}

               {!completed && <button
                  onClick={handleDeleteComment}
                  className="task__button button">Delete comment</button>}
            </div>
         }


      </>

   )
}

export default connect()(Comment)