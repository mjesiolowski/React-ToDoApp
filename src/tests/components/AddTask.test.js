import React from 'react'
import { shallow } from 'enzyme'
import { AddTask } from '../../components/AddTask'
import tasks from '../fixtures/tasks'
import alerts from '../fixtures/alerts'
import moment from 'moment'

let wrapper, dispatch

beforeEach(() => {
  dispatch = jest.fn()

  wrapper = shallow(<AddTask
    dispatch={dispatch}
    alerts={alerts}
    tasks={tasks}
  />)
})

test('should render', () => {
  expect(wrapper).toMatchSnapshot()
})

test('should change taskName state', () => {
  const value = 'test'
  wrapper.find('input').at(0).simulate('change', {
    target: {
      value,
    },
  });

  expect(wrapper.find('input').at(0).prop('value')).toEqual(value)
})

test('should change taskPriority state', () => {
  const value = 'true'
  wrapper.find('select').simulate('change', {
    target: {
      value,
    },
  });

  expect(wrapper.find('select').prop('value')).toEqual(true)
})

test('should change taskDeadline state', () => {
  const value = 20
  wrapper.find('input').at(1).simulate('change', {
    target: {
      value,
    },
  });

  expect(wrapper.find('input').at(1).prop('value')).toEqual(value)
})

test('should handle adding task', () => {
  const task = {
    name: 'dsdsd',
    isPriority: 'true',
    deadline: moment().add(1, 'year').valueOf(),
  }

  wrapper.find('input').at(0).simulate('change', {
    target: {
      value: task.name
    },
  });
  wrapper.find('select').simulate('change', {
    target: {
      value: task.isPriority
    },
  });
  wrapper.find('input').at(1).simulate('change', {
    target: {
      value: task.deadline
    },
  });

  wrapper.find('form').simulate('submit', {
    preventDefault: () => { },
    dateFormat: "DD.MM.YYYY"
  })
  expect(dispatch).toHaveBeenLastCalledWith(
    {
      type: "ADD_TASK",
      task: {
        completed: false,
        completedAt: null,
        createdAt: moment().valueOf(),
        comments: [],
        name: task.name,
        isPriority: true,
        deadline: task.deadline,
        id: expect.any(String)
      }

    }
  )
  expect(wrapper.find('input').at(0).prop('value')).toEqual('')
})

