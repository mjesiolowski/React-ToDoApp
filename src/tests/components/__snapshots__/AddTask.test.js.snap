// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render 1`] = `
<section
  className="addTask"
>
  <h2
    className="title__sub addTask__title"
  >
    AddTask
  </h2>
  <form
    className="form"
    onSubmit={[Function]}
  >
    <div
      className="form__wrapper"
    >
      <label
        className="form__label"
        htmlFor="taskName"
      >
        Add task:
      </label>
      <input
        className="form__input"
        id="taskName"
        onChange={[Function]}
        placeholder="Add task"
        type="text"
        value=""
      />
    </div>
    <div
      className="form__wrapper"
    >
      <label
        className="form__label"
        htmlFor="taskDeadline"
      >
        Deadline:
      </label>
      <input
        className="form__input"
        id="taskDeadline"
        onChange={[Function]}
        value="01.01.1970"
      />
    </div>
    <div
      className="form__wrapper"
    >
      <label
        className="form__label"
        htmlFor="priority"
      >
        Priority: 
      </label>
      <select
        className="form__select"
        id="priority"
        onChange={[Function]}
        value={false}
      >
        <option
          value="false"
        >
          Not important
        </option>
        <option
          value="true"
        >
          Important
        </option>
      </select>
    </div>
    <button
      aria-label="add task"
      className="addTask__button button"
    >
      <i
        className="fas fa-plus-circle"
      />
    </button>
  </form>
</section>
`;
